File: 백준/Silver/11660. 구간 합 구하기 5/구간 합 구하기 5.py
import sys

input = lambda: sys.stdin.readline().rstrip()

N, M = map(int, input().split())
board = [list(map(int, input().split())) for _ in range(N)]

acc_board = [[0] * (N + 1) for _ in range(N + 1)]
acc_board[0][0] = board[0][0]
for i in range(N):
  for j in range(N - 1):
    board[i][j + 1] += board[i][j]
for i in range(N - 1):
  for j in range(N):
    board[i + 1][j] += board[i][j]

# print(board)
for _ in range(M):
  x1, y1, x2, y2 = map(int, input().split())
  s = board[x2 - 1][y2 - 1]
  if x1 - 2 >= 0 and y1 - 2 >= 0:
    s += board[x1 - 2][y1 - 2]
  if x1 - 2 >= 0 and y2 - 1 >= 0:
    s -= board[x1 - 2][y2 - 1]
  if x2 - 1 >= 0 and y1 - 2 >= 0:
    s -= board[x2 - 1][y1 - 2]
  print(s)

File: 백준/Silver/11660. 구간 합 구하기 5/구간 합 구하기 5.java
import java.io.*;
import java.util.*;
public class Main {
	static int N, M;
	static int[][] board;
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		String[] inputs = br.readLine().split(" ");
		N = Integer.parseInt(inputs[0]);
		M = Integer.parseInt(inputs[1]);
		board = new int[N][N];
		StringTokenizer st;
		for (int n = 0; n < N; n++) {
			st = new StringTokenizer(br.readLine());
			for (int i = 0; i < N; i++) {
				board[n][i] = Integer.parseInt(st.nextToken());
			}
		}
		for (int i = 0; i < N; i++) {
			for (int j = 1; j < N; j++) {
				board[i][j] += board[i][j - 1];
			}
		}
		for (int i = 0; i < N; i++) {
			for (int j = 1; j < N; j++) {
				board[j][i] += board[j-1][i];
			}
		}
		for (int m = 0; m < M; m++) {
			inputs = br.readLine().split(" ");
			int x1 = Integer.parseInt(inputs[0]) - 1;
			int y1 = Integer.parseInt(inputs[1]) - 1;
			int x2 = Integer.parseInt(inputs[2]) - 1;
			int y2 = Integer.parseInt(inputs[3]) - 1;
			int ans = board[x2][y2];
			if (x1 > 0) ans -= board[x1 - 1][y2];
			if (y1 > 0) ans -= board[x2][y1 - 1];
			if (x1 > 0 && y1 > 0)
				ans += board[x1 - 1][y1 - 1];
			System.out.println(ans);
		}

	}
}
